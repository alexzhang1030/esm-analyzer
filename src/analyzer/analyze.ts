import type { TreeNode } from 'to-path-tree'
import { getExportByName, getImportByName } from '..'
import type { TreeNodeData, TreeNodeItem } from '.'
import { rangeLoop } from '@/utils'

const ENTRY = 'index'

function findTheEntry(node: TreeNodeItem) {
  if (node.filename === ENTRY)
    return node
  return node.parent.items.find(item => item.filename === ENTRY) ?? null
}

function findTheParentEntry(node: TreeNode<TreeNodeData>) {
  return node.items.find(item => item.filename === ENTRY) ?? null
}

function findSibling(node: TreeNodeItem, name: string) {
  return node.parent.items.find(item => item.filename === name) ?? null
}

function findChildren(node: TreeNode<TreeNodeData>, name: string) {
  return node.items.find(item => item.filename === name) ?? null
}

export function isNotPath(path: string) {
  return /^(?!\.\/|\.|\.\.\/).*/.test(path)
}

function getParentPaths(path: string) {
  const parentPaths = path.split('/').filter(item => item === '..')
  return parentPaths.length
}

function processParentPath(path: string, node: TreeNodeItem) {
  const parentLength = getParentPaths(path)
  if (parentLength === 0)
    return null
  let result: TreeNode<TreeNodeData> | null = null
  rangeLoop(parentLength, () => {
    result = result ? result.parent! : node.parent
  })
  return result
}

function getLastImport(path: string) {
  return /\/([^/]+)$/.exec(path)?.[1] ?? null
}

// ../foo/bar
export function getTargetNodeByPath(path: string, node: TreeNodeItem) {
  if (path === '.')
    return findTheEntry(node)
  if (path === '..')
    return findTheParentEntry(node.parent)
  if (isNotPath(path)) {
    // TODO: resolve: is import from node_modules
    return null
  }
  // ./foo or ./../foo
  if (path.startsWith('./')) {
    path = path.slice(2)
    if (isNotPath(path))
      return findSibling(node, path)
    if (path.startsWith('../')) {
      const resultNode = processParentPath(path, node)
      if (!resultNode)
        return null
      const importName = getLastImport(path)
      if (!importName)
        return null
      return findChildren(resultNode, importName)
    }
  }
  // ../foo or ../../foo
  if (path.startsWith('../')) {
    const resultNode = processParentPath(path, node)
    if (!resultNode)
      return null
    const importName = getLastImport(path)
    if (!importName)
      return null
    return findChildren(resultNode, importName)
  }
}

export class Analyzer {
  #node: TreeNodeItem

  constructor(node: TreeNodeItem) {
    this.#node = node
  }

  findImportLocation(name: string) {
    // 1. find all import statements
    const imports = this.#node.data?.scan.imports
    if (!imports)
      return null
    // 2. get target node
    const source = getImportByName(name, imports)?.source
    if (!source)
      return null
    // 3. get target node by path
    const targetNode = getTargetNodeByPath(source, this.#node)
    // 4. get export statement
    if (!targetNode)
      return null
    const exportStmt = targetNode.data!.scan.exports
    const exportNode = getExportByName(name, exportStmt)
    if (!exportNode)
      return null
    return {
      nodeData: exportNode,
      file: targetNode.path,
    }
  }
}
