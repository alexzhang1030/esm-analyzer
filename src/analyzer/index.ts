import mitt from 'mitt'
import pLimit from 'p-limit'
import type { NodeItem } from 'to-path-tree'
import { pathToTree, walkPathTree } from 'to-path-tree'
import type { AcceptableLang, ScanExportResult, ScanImportResult, ScanVariableDeclarationResult } from '..'
import { scan } from '..'
import { getLangByFileName, isAcceptableLang } from '@/common'
import { loop } from '@/utils'

const limit = pLimit(10)

export interface TreeNodeData {
  code: string
  scan: {
    imports: ScanImportResult[]
    exports: ScanExportResult[]
    variables: ScanVariableDeclarationResult[]
  }
}

export type TreeNodeItem = NodeItem<TreeNodeData>

interface MapData {
  source: {
    code: string
    lang: AcceptableLang
  }
  referToNode?: TreeNodeItem
}

export class Project {
  #name: string
  #mapping: Map</* _absolute_ file name */string, MapData> = new Map()

  #progress = 0
  #mitt = mitt<{
    progress: number
  }>()

  get name() {
    return this.#name
  }

  constructor(name: string) {
    this.#name = name
  }

  #emitProgress(progress: number) {
    this.#progress = progress
    this.#mitt.emit('progress', progress)
  }

  #scanFile(code: string, lang: AcceptableLang) {
    return scan(code, lang)
  }

  /**
   *
   * @param fileName must be **absolute** file name
   * @param fileCode
   */
  addFile(fileName: string, fileCode: string) {
    if (!isAcceptableLang(fileName))
      throw new Error(`File name "${fileName}" is not acceptable.`)
    const lang = getLangByFileName(fileName)
    this.#mapping.set(fileName, {
      source: {
        code: fileCode,
        lang,
      },
    })
  }

  onProgress(callback: (progress: number) => void) {
    this.#mitt.on('progress', callback)
  }

  async prepare() {
    const codes = this.#mapping.entries()
    const filepaths = Array.from(codes, ([filename]) => filename)
    const tasks: (() => Promise<void>)[] = []
    const tree = pathToTree<TreeNodeData>(filepaths, {
      getData: (node) => {
        return {
          code: this.#mapping.get(node.path)!.source.code,
          scan: {} as TreeNodeData['scan'],
        }
      },
    })
    walkPathTree(tree, (node) => {
      loop(node.items, (item) => {
        const { data, ext } = item
        const { code } = data!
        tasks.push(limit(() => {
          const r = this.#scanFile(code, ext as AcceptableLang)
          this.#emitProgress(this.#progress + 1)
          item.data!.scan = r
          this.#mapping.get(item.path)!.referToNode = item
        }) as any)
      })
    })
    await Promise.all(tasks)
  }

  getTreeNode(filename: string) {
    return this.#mapping.get(filename)!.referToNode
  }
}
