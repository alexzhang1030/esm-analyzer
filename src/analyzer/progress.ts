import mitt from 'mitt'

export class Progress {
  #totalTasksCount: number
  #progress = 0
  #emitter = mitt<{
    progress: number
  }> ()

  constructor(count?: number) {
    this.#totalTasksCount = count ?? 0
  }

  #calcProgress() {
    return Math.round(this.#progress / this.#totalTasksCount).toFixed(2)
  }

  onProgress(callback: (progress: number) => void) {
    this.#emitter.on('progress', callback)
  }

  reduce() {
    this.#progress++
    this.#emitter.emit('progress', Number(this.#calcProgress()))
  }

  addProgress(count: number) {
    this.#progress += count
  }
}
