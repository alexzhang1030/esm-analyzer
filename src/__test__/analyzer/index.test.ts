import { Project } from '@/analyzer'
import { Analyzer } from '@/analyzer/analyze'

describe('project', async () => {
  test('file paths to tree', () => {

  })
  const code1 = {
    filename: '/src/bar.js',
    code: `
      export const bar = 'bar'
    `,
  }
  const code2 = {
    filename: '/src/foo.js',
    code: `
      import { bar } from './bar'
      export const foo = bar
    `,
  }
  const p = new Project('test')
  p.addFile(code1.filename, code1.code)
  p.addFile(code2.filename, code2.code)
  await p.prepare()
  test('find the location of bar that /src/foo.js imported', async () => {
    const node = p.getTreeNode(code2.filename)
    const a = new Analyzer(node!)
    const { file, nodeData } = a.findImportLocation('bar')!
    expect({ file, loc: nodeData.loc }).toMatchInlineSnapshot(`
      {
        "file": "/src/bar.js",
        "loc": {
          "end": {
            "column": 30,
            "index": 31,
            "line": 2,
          },
          "start": {
            "column": 6,
            "index": 7,
            "line": 2,
          },
        },
      }
    `)
  })
})
