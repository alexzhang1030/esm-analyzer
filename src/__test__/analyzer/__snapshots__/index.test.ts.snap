// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`project > find the location of bar 1`] = `
{
  "end": {
    "column": 30,
    "index": 31,
    "line": 2,
  },
  "start": {
    "column": 6,
    "index": 7,
    "line": 2,
  },
}
`;
